Sprint 10

superaydumov opened the pull request:
Pull request according to Yandex Practikum course.

superaydumov committed:
Created folders for view controllers.
Added "Constants" to the project (Unsplash keys, redirect information, api URL etc).

superaydumov committed:
Added AuthViewController, created individual view controller for it and set up UI elements (buttons, background, logo).

superaydumov committed:
Added WebViewViewController and it's logic (view controller works with Unsplash API webView).
Modified AuthViewController (added prepare method to get information about user actions, made extension to follow WebViewViewControllerDelegate).

superaydumov committed:
Added progress view to WebViewViewController to show Unsplash login page download progress.
Added methods to refresh progress view using KVO while web page is loading.

superaydumov committed:
Added logic for app to do network requests and responses (methods, properties and extensions) and decode received data.

superaydumov committed:
Refactored code, added separate files for protocols.

superaydumov committed:
Added method to center image in SingleImageViewController after scrolling.

superaydumov committed:
Refactored code to delete mistakes.
Added SplashviewController to navigate user through different flows if user had already logged in or not.

protos17 commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,54 @@
+        loginButtonUISetup()
+    }
+    
+    func loginButtonUISetup() {
```
Здесь стоит добавить `private`
Необходимо указывать private  для того, чтобы спрятать детали реализации конкретной части функциональности, когда они используются только внутри области объявления.

protos17 commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,54 @@
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == webSegueIdentifier {
+            guard let webViewViewController = segue.destination as? WebViewViewController else {
+                fatalError("Failed to prepare for \(webSegueIdentifier)") }
```
Хочу отметить, что в продакшене не используется fatalError, кроме случаев, когда это необходимо, например, в банковской приложении при обнаружении джейлбрейка.

protos17 commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,54 @@
+
+final class AuthViewController: UIViewController {
+    
+    private let webSegueIdentifier = "ShowWebView"
+    
+    weak var delegate: AuthViewControllerDelegate?
+    
+    @IBOutlet weak var loginButton: UIButton!
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
```
Обычно используем такой порядок: сначала перечисляем свойства класса, затем функции. При этом также делим на приватные и публичные блоки. А функции перечисляют в таком порядке: `init`ы, `override`ы, блоки публичных и приватных функций, и далее блок c аннотацией `@IBAction`

protos17 commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,54 @@
+    func webViewViewControllerDidCancel(_vc: WebViewViewController) {
+        dismiss(animated: true)
+    }
+    
+    
```
Лишние отступы рекомендую убирать

protos17 commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,103 @@
+import Foundation
+import WebKit
+
+fileprivate let UnsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Рекомендую перенести в Constants

protos17 commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,103 @@
+import WebKit
+
+fileprivate let UnsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
+fileprivate let constants = Constants()
```
Константы лучше сделать как static let и обращаться к ним сразу через точку:

Constants.accessKey

protos17 commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,54 @@
+    
+    weak var delegate: AuthViewControllerDelegate?
+    
+    @IBOutlet weak var loginButton: UIButton!
```
Рекомендую указать как private

protos17 commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,103 @@
+
+final class WebViewViewController: UIViewController {
+    
+    @IBOutlet weak var webView: WKWebView!
+    @IBOutlet weak var progressView: UIProgressView!
```
Рекомендую указать как private

protos17 commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,103 @@
+            URLQueryItem(name: "response_type", value: "code"),
+            URLQueryItem(name: "scope", value: constants.accessScope)
+        ]
+        let url = urlComponents.url!
```
Рекомендую сделать проверку, вместо принудительного извлечения

protos17 commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,103 @@
+        progressView.isHidden = fabs(webView.estimatedProgress - 1.0) <= 0.0001
+    }
+    
+    @IBAction func didTapBackButton(_ sender: UIButton) {
```
Рекомендую указать как private

protos17 commented on ImageFeed/Constants.swift:
```diff
@@ -0,0 +1,17 @@
+import Foundation
+
+struct Constants {
+    let accessKey = "5cknYblskv1f9HgymUwDSz7X5dGqM9-0f1mpmVzebqU"
+    let secretKey = "CH3eNm3fTDWw60be512fQT8usbv6Gfw4Z7jsYnJv9z8"
+    let redirectURI = "urn:ietf:wg:oauth:2.0:oob"
+    
+    let accessScope = "public+read_user+write_likes"
+    let defaultbaseURL = URL (string: "https://api.unsplash.com/")!
```
Рекомендую указать как static let и образаться без создания экземпляра структуры

protos17 commented on ImageFeed/Models/OAuthTokenResponseBody.swift:
```diff
@@ -0,0 +1,22 @@
+    private enum CodingKeys: String, CodingKey {
+        case accessToken = "access_token"
+        case tokenType = "token_type"
+        case scope = "scope"
```
Если CodingKey совпадает с полем, его можно не писать:

case tokenType = "token_type"
case scope

protos17 commented on ImageFeed/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,144 @@
+    
+    // MARK: - Properties
+    
+    static let shared = OAuth2Service()
```
При использовании синглтона рекомендую еще добавлять:

private init() {}

protos17 commented on ImageFeed/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,144 @@
+            guard let self = self else { return }
+            switch result {
+            case .success (let body):
+                let authToken = body.accessToken
+                self.authToken = authToken
```
Можно присвоить сразу, без введения новой константы:

self.authToken = body.accessToken

superaydumov replied:
Не поручилось использовать эту конструкцию. Выдает ошибку, а как её правильно решить не знаю (что поставить во второе условие nil coalescing operator).

protos17 replied:
Для таких случаев рекомендую обращаться к Наставнику или Наставнице :)

protos17 commented on ImageFeed/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,144 @@
+            + "&&code=\(code)"
+            + "&&grant_type=authorization_code",
+            httpMethod: "POST",
+            baseURL: URL(string: "https://unsplash.com")!)
```
Реокмендую перенести в Constants

protos17 commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,78 @@
+            guard
+                let navigationController = segue.destination as? UINavigationController,
+                let viewCotroller = navigationController.viewControllers[0] as? AuthViewController
+            else { fatalError("Failed to prepare for \(ShowAuthentificationScreenSegueIdentifier)")}
```
Рекомендую fatalError заменить на return

superaydumov replied:
при попытке заменить на return и изменить функцию для возвращения строки, сразу ругается на то, что функция не переопределяет метод из superclass

superaydumov replied:
<img width="838" alt="image" src="https://github.com/superaydumov/ImageFeed/assets/131291379/4a08d7fb-83f1-4e7e-8e92-2166fcb4c309">

protos17 replied:
Ну правильно, не нужно сам метод менять, нужно всего лишь заменить на return:
guard ... as? AuthViewController else { return }

protos17 commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,78 @@
+            case .success:
+                self.switchToTabBarController()
+            case .failure:
+                print ("There's a problem with token!")
```
Рекомендую показать алерт с ошибкой пользователю

superaydumov committed:
Refactored code after reviewer comments, added marks everywhere, "private" identifications where needed etc.
Changed ShowAuthenticationScreen's Presentation to FullScreen mode in Main.
Changed Constants structure (added "static" to constants for easy access), added new properties (moved from other files).
Changed OAuthTokenResponseBody structure (removed enum CodingKeys) to use keyDecodingStrategy (convert snake case to camel case while decoding).
Added private init() {} to static shared property to match singletone pattern.
Added light status bars where needed.
